prover/src/matrix/segments.rs

```diff
@@ -0,0 +1,495 @@
+        self.data[i][3] = temp[3] + self.data[j][3];
+        self.data[j][3] = temp[3] - self.data[j][3];
+
+        // apply of index 4 of twiddle.
+        self.data[j][4] *= twiddle;
+        self.data[i][4] = temp[4] + self.data[j][4];
+        self.data[j][4] = temp[4] - self.data[j][4];
+
+        // apply of index 5 of twiddle.
+        self.data[j][5] *= twiddle;
+        self.data[i][5] = temp[5] + self.data[j][5];
+        self.data[j][5] = temp[5] - self.data[j][5];
+
+        // apply of index 6 of twiddle.
+        self.data[j][6] *= twiddle;
+        self.data[i][6] = temp[6] + self.data[j][6];
+        self.data[j][6] = temp[6] - self.data[j][6];
+
+        // apply of index 7 of twiddle.
+        self.data[j][7] *= twiddle;
+        self.data[i][7] = temp[7] + self.data[j][7];
+        self.data[j][7] = temp[7] - self.data[j][7];
+    }
+
+    fn swap(&mut self, i: usize, j: usize) {
+        self.data.swap(i, j);
+    }
+
+    fn shift_by_series(&mut self, offset: E::BaseField, increment: E::BaseField) {
+        let increment = E::from(increment);
+        let mut offset = E::from(offset);
+        for row_idx in 0..self.len() {
+            // apply on index 0.
+            self.data[row_idx][0] *= offset;
+
+            // apply on index 1.
+            self.data[row_idx][1] *= offset;
+
+            // apply on index 2.
+            self.data[row_idx][2] *= offset;
+
+            // apply on index 3.
+            self.data[row_idx][3] *= offset;
+
+            // apply on index 4.
+            self.data[row_idx][4] *= offset;
+
+            // apply on index 5.
+            self.data[row_idx][5] *= offset;
+
+            // apply on index 6.
+            self.data[row_idx][6] *= offset;
+
+            // apply on index 7.
+            self.data[row_idx][7] *= offset;
+
+            offset *= increment;
+        }
+    }
+
+    fn shift_by(&mut self, offset: E::BaseField) {
+        let offset = E::from(offset);
+
+        for row_idx in 0..self.len() {
+            // apply on index 0.
+            self.data[row_idx][0] *= offset;
+
+            // apply on index 1.
+            self.data[row_idx][1] *= offset;
+
+            // apply on index 2.
+            self.data[row_idx][2] *= offset;
+
+            // apply on index 3.
+            self.data[row_idx][3] *= offset;
+
+            // apply on index 4.
+            self.data[row_idx][4] *= offset;
+
+            // apply on index 5.
+            self.data[row_idx][5] *= offset;
+
+            // apply on index 6.
+            self.data[row_idx][6] *= offset;
+
+            // apply on index 7.
+            self.data[row_idx][7] *= offset;
+        }
+    }
+}
+
+// SEGMENTS
+// ================================================================================================
+
+/// Represents a collection of Segment objects.
+#[derive(Debug, Clone)]
+pub struct Segments<E>
+where
+    E: FieldElement,
+{
+    matrix: Vec<Segment<E>>,
+}
```

Comment:
Why do we need this `Segments` struct? From what I can tell, it is used only for implementing the iterators, but then the iterators don't seem to do anything special (unless I missed something). Could we just use `Vec<Segment>` instead of creating a special struct?

Reply:
Yeah- we don't need it in the current implementation. I've removed it in the latest update.